### PR TITLE
NEAR: header verification and state update functions

### PR DESCRIPTION
**Pull request by blasrodri**

### CODE REVIEW

**Comment by seunlanlege** [comment]
can we close the other PRs?

File: modules/Cargo.toml [hunk 1]
```diff
@@ -40,6 +40,8 @@ clock = ["tendermint/clock", "time/std"]
 # Depends on the `testgen` suite for generating Tendermint light blocks.
 mocks = ["tendermint-testgen", "clock", "std", "sha3", "ripemd", "ics11_beefy"]
 ics11_beefy = ["sp-io", "sp-core", "sp-std", "beefy-primitives", "sp-runtime", "sp-trie", "pallet-mmr-primitives", "beefy-client", "beefy-client-primitives"]
+ics13_near = ["skip_host_consensus_verification"]
+skip_host_consensus_verification = []
```
Comment (seunlanlege) [inline modules/Cargo.toml]: unneccesary

File: modules/src/clients/host_functions.rs [hunk 2]
```diff
@@ -1,10 +1,18 @@
...
+use near_primitives_wasm::HostFunctions as MaybeNearHostFunctions;
+
+#[cfg(not(any(test, feature = "ics13_near")))]
+trait MaybeNearHostFunctions {}
 
 /// This trait captures all the functions that the host chain should provide for
 /// crypto operations.
-pub trait HostFunctionsProvider: Clone + Send + Sync + Default {
+pub trait HostFunctionsProvider:
+    Clone + Send + Sync + Default + BeefyHostFunctions + MaybeNearHostFunctions
```
Comment (seunlanlege) [inline modules/src/clients/host_functions.rs]: `BeefyHostFunctions` is already implemented for `HostFunctionsManager`

Comment (seunlanlege) [inline modules/src/clients/host_functions.rs]:
```suggestion
    Clone + Send + Sync + Default 
```

see: https://github.com/ComposableFi/ibc-rs/blob/55fc39714c150a11c1d13a93dd6de60eb051d47c/modules/src/clients/host_functions.rs#L77-L81

File: modules/src/clients/ics13_near/client_def.rs [hunk 3]
```diff
@@ -56,31 +58,79 @@ impl<T: HostFunctionsProvider> ClientDef for NearClient<T> {
...
         _ctx: &dyn ReaderContext,
         _client_id: ClientId,
-        client_state: Self::ClientState,
+        mut client_state: Self::ClientState,
         header: Self::Header,
     ) -> Result<(), Error> {
-        // your light client, shouldn't do storage anymore, it should just do verification here.
-        validate_light_block::<T>(&header, client_state)
+        // TODO: optimize the verification, according to https://nomicon.io/ChainSpec/LightClient
+        for header in header.inner {
```
Comment (seunlanlege) [inline modules/src/clients/ics13_near/client_def.rs:65]: this isn't using batch proofs via the block merkle root. Thought this was already implemented?

File: modules/src/clients/ics13_near/client_state.rs [hunk 4]
```diff
@@ -1,48 +1,38 @@
...
-    head: LightClientBlockView,
+    pub head: LightClientBlockView,
     current_epoch: CryptoHash,
+    pub epoch_num: u64,
+    pub prev_epoch_num: u64,
     next_epoch: CryptoHash,
-    current_validators: Vec<ValidatorStakeView>,
+    pub epoch_block_producers: NearBlockProducers,
     next_validators: Vec<ValidatorStakeView>,
+    pub prev_lite_block: LightClientBlockView,
```
Comment (seunlanlege) [inline modules/src/clients/ics13_near/client_state.rs]: probably best to make everything here public

File: modules/src/clients/ics13_near/client_state.rs [hunk 5]
```diff
@@ -53,12 +43,15 @@ impl ClientState for NearClientState {
...
+    }
+
+    fn is_frozen(&self) -> bool {
+        self.frozen_height().is_some()
     }
 
     fn frozen_height(&self) -> Option<crate::Height> {
-        // TODO: validate this
-        Some(self.head.get_height())
+        todo!()
```
Comment (seunlanlege) [inline modules/src/clients/ics13_near/client_state.rs]:
```suggestion
        self.frozen_height.clone()
```

**Review by seunlanlege — CHANGES_REQUESTED** [review]
Looks good, just needs to address comments and the numerous todo's

File: modules/src/clients/ics13_near/client_def.rs (hunk 3 above)
Comment (vmarkushin) [reply modules/src/clients/ics13_near/client_def.rs:65]: I decided to postpone it for now, just to make sure everything works without them. Should I work on it in this PR?

Comment (seunlanlege) [reply modules/src/clients/ics13_near/client_def.rs:65]: no need, lets just create a tracking issue. So we can get to integration testing quickly.

Comment (vmarkushin) [reply modules/src/clients/ics13_near/client_def.rs:65]: Sounds good

**Comment by seunlanlege** [comment]
we're no longer maintaining clients in this repo, stay tuned pls

**Pull request closed without merging**